csrc/multidevice/utils.cpp

```diff
@@ -464,6 +471,27 @@ void insertShardedAxisReordering(Fusion* fusion) {
   }
 }
 
+void setShardedAllocationDomain(Fusion* fusion) {
+  for (auto expr : fusion->exprs()) {
+    if (isResharding(expr)) {
+      for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        for (auto c : tv->getContiguity()) {
+          if (c.has_value()) {
+            NVF_CHECK(
+                c.value(),
+                "Resharding expression input must be contiguous ",
+                expr);
+          }
+        }
+        setShardedAllocationDomain(tv);
```

Comment:
Can somebody please explain why this is necessary? IIUC, we confirm this tensor is contiguous. Isn't that sufficient?

Reply:
So, this is not necessary if the allocation order inference is not done?

Reply:
I think so.

For the record, it's not just allocation order inference, alias passes could also update allocation domain introducing similar issue. See Jingyue's comment in https://github.com/NVIDIA/Fuser/pull/2245#issuecomment-2111317055

cc'ing @cowanmeg for sanity check